Sources/Stores/Store.swift

```diff
@@ -119,77 +95,101 @@ public extension Store {
-                do {
-                    let data = try inner()
+        // 1st - Fetch data from the Persistence
+        self.getPersistedData(for: resource) { (data) in
 
-                    guard cancelable.isCancelled == false else { return completion(nil, .cancelled, false) }
+            // If we have data we don't need to go to the network
+            if let data = data {
 
-                    // parse the new value from the data
-                    let value = try resource.parser(data)
+                // parse the new value from the data
+                self.process(data, fromCache: true, resource: resource, cancelable: cancelable, completion)
 
+            } else {
+
+                // 2nd - Try to fetch Data from Network
+                cancelable.networkCancelable = self.getNetworkData(resource) { (data, error) in
+
+                    // Check if it's cancelled
                     guard cancelable.isCancelled == false else { return completion(nil, .cancelled, false) }
 
-                    // update persistence with new value
-                    self.persist(data, for: resource)
-
-                    completion(value, nil, false)
-                } catch let Network.Error.url(error as NSError) where error.domain == NSURLErrorDomain
-                                                                   && error.code == NSURLErrorCancelled {
-                    completion(nil, .cancelled, false)
-                } catch let error as Network.Error {
-                    completion(nil, .network(error), false)
-                } catch let error as Parse.Error {
-                    completion(nil, .parse(error), false)
-                } catch {
-                    completion(nil, .other(error), false)
+                    // If we got an error we need to check if we have the data on disk
+                    if let error = error {
+
+                        return completion(nil, error, false)
+
+                    } else if let data = data {
+
+                        // parse the new value from the data
+                        self.process(data, fromCache: false, resource: resource, cancelable: cancelable, completion)
+                    }
                 }
             }
         }
 
-        // check persistence to see if we have a hit and return immediately if so
-        getPersistedData(for: resource) {
-            guard let data = $0 else {
-                return fetch(resource: resource)
-            }
+        return cancelable
+    }
 
-            do {
-                let value = try resource.parser(data)
-                completion(value, nil, true)
-            } catch {
-                // try to fetch fresh data if parsing of existent data failed
-                // TODO: remove from persistence?
-                return fetch(resource: resource)
+    // MARK: Processing Methods
+
+    private func process<Resource: NetworkResource & PersistableResource>(_ data: Data,
```

Comment:
this `process` is essentially a `parseAndPersist`, right? Not sure if you agree, but at least we would have something a bit more specific than just `process` 😆